source/postcard-rpc/src/server/impls/embedded_io_async_v0_6.rs

```diff
@@ -0,0 +1,302 @@
+        let used = body_to_flavor(msg, flavor)?;
+
+        // Write it all to the serial port now
+        t.write_all(used)
+            .await
+            .map_err(|_| WireTxErrorKind::ConnectionClosed)?;
+
+        // defmt::println!("SENT NORMAL {=usize} {=[u8]}", used.len(), used);
+        // We did it! yaaaay!
+        Ok(())
+    }
+
+    async fn send_raw(&self, buf: &[u8]) -> Result<(), Self::Error> {
+        let mut guard = self.t.lock().await;
+        let EioWireTxInner { t, tx_buf, .. } = guard.deref_mut();
+
+        // Create a cobs-encoding flavor using our temp buffer
+        let mut flavor = flava_flav(tx_buf)?;
+
+        flavor.try_extend(buf).map_err(|_| WireTxErrorKind::Other)?;
+        let used = flavor.finalize().map_err(|_| WireTxErrorKind::Other)?;
+
+        // Write it all to the serial port now
+        t.write_all(used)
+            .await
+            .map_err(|_| WireTxErrorKind::ConnectionClosed)?;
+
+        // defmt::println!("SENT RAW");
+        // We did it! yaaaay!
+        Ok(())
+    }
+
+    async fn send_log_str(&self, kkind: VarKeyKind, s: &str) -> Result<(), Self::Error> {
+        let mut guard = self.t.lock().await;
+        let EioWireTxInner { t, tx_buf, log_seq } = guard.deref_mut();
+
+        // Create a cobs-encoding flavor using our temp buffer
+        let mut flavor = flava_flav(tx_buf)?;
+
+        let key = match kkind {
+            VarKeyKind::Key1 => VarKey::Key1(LoggingTopic::TOPIC_KEY1),
+            VarKeyKind::Key2 => VarKey::Key2(LoggingTopic::TOPIC_KEY2),
+            VarKeyKind::Key4 => VarKey::Key4(LoggingTopic::TOPIC_KEY4),
+            VarKeyKind::Key8 => VarKey::Key8(LoggingTopic::TOPIC_KEY),
+        };
+        let ctr = *log_seq;
+        *log_seq = log_seq.wrapping_add(1);
+        let wh = VarHeader {
+            key,
+            seq_no: VarSeq::Seq2(ctr),
+        };
+
+        header_to_flavor(&wh, &mut flavor)?;
+        let used = body_to_flavor(s, flavor)?;
+
+        // Write it all to the serial port now
+        t.write_all(used)
+            .await
+            .map_err(|_| WireTxErrorKind::ConnectionClosed)?;
+
+        // We did it! yaaaay!
+        Ok(())
+    }
+
+    async fn send_log_fmt<'a>(
+        &self,
+        _kkind: VarKeyKind,
+        _a: Arguments<'a>,
+    ) -> Result<(), Self::Error> {
+        todo!()
+    }
+}
+
+/// ...
+pub struct EioWireRx<R: Read> {
+    /// ...
+    pub remain: &'static mut [u8],
+    /// ...
+    pub offset: usize,
+    /// ...
+    pub rx: R,
+}
+
+fn copy_backwards(buf: &mut [u8], start: usize) {
+    if buf.len() < start {
+        return;
+    }
+    let count = buf.len() - start;
+    let base = buf.as_mut_ptr();
+    unsafe {
+        core::ptr::copy(base.add(start).cast_const(), base, count);
+    }
+}
+
+impl<R: Read> WireRx for EioWireRx<R> {
+    type Error = WireRxErrorKind;
+
+    async fn receive<'a>(&mut self, buf: &'a mut [u8]) -> Result<&'a mut [u8], Self::Error> {
```

Comment:
Honestly, I should probably make a version of `CobsAccumulator` that works on borrowed slices, this is a bunch of bespoke code that has room for errors.